rules/onchain_budget.go

```diff
@@ -0,0 +1,694 @@
+package rules
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/lightninglabs/lightning-terminal/firewalldb"
+	"github.com/lightninglabs/lightning-terminal/litrpc"
+	mid "github.com/lightninglabs/lightning-terminal/rpcmiddleware"
+	"github.com/lightninglabs/lightning-terminal/session"
+	"github.com/lightningnetwork/lnd/lnrpc"
+	"google.golang.org/protobuf/proto"
+)
+
+var (
+	// Compile-time checks to ensure that OnChainBudget, OnChainBudgetMgr
+	// and OnChainBudgetEnforcer implement the appropriate Manager, Enforcer
+	// and Values interface.
+	_ Manager  = (*OnChainBudgetMgr)(nil)
+	_ Enforcer = (*OnChainBudgetEnforcer)(nil)
+	_ Values   = (*OnChainBudget)(nil)
+)
+
+const (
+	// OnChainBudgetName is the string identifier of the OnChainBudgetMgr
+	// rule.
+	OnChainBudgetName = "on-chain-budget"
+
+	// spendAmtKey is the key that will be used in the persisted KV store to
+	// store the total amount that has been spent.
+	onChainSpent = "spent-amt"
+
+	// pendingSpentKey is the key that will be used in the persisted KV
+	// store to keep track of the total pending spent amount.
+	onChainPending = "pending-spent"
+)
+
+// OnChainBudgetMgr manages the OnChainBudget rule.
+type OnChainBudgetMgr struct {
+	// The mutex is used to ensure that only one Enforcer created by the
+	// manager can run the HandleRequest and HandleResponse functions at
+	// any given time. This prevents db entry race conditions.
+	sync.Mutex
+}
+
+// Stop cleans up the resources held by the manager.
+//
+// NOTE: This is part of the Manager interface.
+func (o *OnChainBudgetMgr) Stop() error {
+	return nil
+}
```

Comment:
Just noting that I think we should be able to use this function to delay litd's shutdown until we've either received a response from `lnd` for a channel open request, or until we time out.

I also just wanted to check that my understanding is correct that we decided to delay that extra functionality and instead add it in a follow-up PR. Is that correct?

Reply:
I think that would be ideal to have, but would require a bit more work to check that we don't create side effects and I'm not sure how to test this in an automated way (we'd need to deliberately create a delayed call). I'll experiment around with that, but would suggest to keep this optional for this PR, but to be addressed soon.

Reply:
surely pretty easy to test? just extend `TestOnChainBudgetCheckRequest`: call a request, call Stop - see that it waits, then call response & see that Stop exits

Reply:
this would be a bit complicated to actually implement though afaict since what happens if LND stops before lit? we would need to hook in to that to know which requests we can just not wait for